File: leetcode/3300-3400/3312. 查询排序后的最大公约数.py
from collections import *
from functools import *
from math import *
from itertools import *
from heapq import *
from bisect import *
from typing import *
from queue import *
from string import *
from random import *

DIRS = [(1, 1), (1, -1), (-1, -1), (-1, 1)]
Max = lambda a, b: b if b > a else a
Min = lambda a, b: b if b < a else a
INF = float("inf")
MOD = int(1e9 + 7)


class Solution:
    def gcdValues(self, nums: List[int], queries: List[int]) -> List[int]:
        mx = max(nums)
        cnt = [0] * (mx + 1)
        for x in nums:
            cnt[x] += 1

        cnt_gcd = [0] * (mx + 1)
        for i in range(mx, 0, -1):
            c = 0
            for j in range(i, mx + 1, i):
                c += cnt[j]
                cnt_gcd[i] -= cnt_gcd[j]
            cnt_gcd[i] += c * (c - 1) // 2

        for i in range(1, mx + 1):
            cnt_gcd[i] = cnt_gcd[i] + cnt_gcd[i - 1]

        return [bisect_right(cnt_gcd, q) for q in queries]

File: leetcode/3300-3400/3312. 查询排序后的最大公约数.cpp
//
// Created by 123 on 25-7-28.
//
#include "bits/stdc++.h"
using namespace std;
using ll = long long;

class Solution {
public:
	vector<int> gcdValues(vector<int>& nums, vector<long long>& queries) {
        ll n = nums.size();
		ll mx = ranges::max(nums);

		vector<ll> cnt(mx + 1,0 );
		for (ll x : nums) {
			cnt[x] += 1;
		}

		vector<ll> dp(mx + 1, 0);
		for (ll i = mx; i >= 1; i--) {
			ll t = 0;
			for (ll j = i; j <= mx; j+= i) {
				t += cnt[j];
				dp[i] -= dp[j];
			}
			dp[i] += t * (t - 1) / 2;
		}

		for (ll i = 1; i <= mx; i++) {
			dp[i] += dp[i - 1];
		}

		vector<int> ans(queries.size());
		for (ll i = 0; i < queries.size(); i++) {
			ll x = queries[i];
			ans[i] = ranges::upper_bound(dp, x) - dp.begin();
		}
		return ans;
	}
};